packages/test/test-end-to-end-tests/src/test/channelSwap.spec.ts

```diff
@@ -0,0 +1,219 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+import { describeNoCompat } from "@fluid-internal/test-version-utils";
+import { IContainer, LoaderHeader } from "@fluidframework/container-definitions";
+import {
+	ITestObjectProvider,
+	createSummarizerFromFactory,
+	summarizeNow,
+} from "@fluidframework/test-utils";
+import {
+	DataObjectTypes,
+	DataObject,
+	IDataObjectProps,
+	DataObjectFactory,
+	ContainerRuntimeFactoryWithDefaultDataStore,
+} from "@fluidframework/aqueduct";
+import { IChannel, IChannelFactory } from "@fluidframework/datastore-definitions";
+import { ISharedDirectory } from "@fluidframework/map";
+import { IFluidHandle } from "@fluidframework/core-interfaces";
+import { requestFluidObject } from "@fluidframework/runtime-utils";
+import { SharedTree } from "@fluid-experimental/tree";
+import {
+	SharedTreeFactory as NewSharedTreeFactory,
+	ISharedTree as ISharedTree2,
+} from "@fluid-experimental/tree2";
+import { IContainerRuntimeOptions, ContainerRuntime } from "@fluidframework/container-runtime";
+import { UnsafeHotSwapFluidDataStoreRuntime } from "@fluid-experimental/datastore-hot-swap";
+
+export class MigratorDataObject<I extends DataObjectTypes = DataObjectTypes> extends DataObject<I> {
+	private readonly hotSwapRuntime: UnsafeHotSwapFluidDataStoreRuntime;
+
+	public get _root(): ISharedDirectory {
+		return this.root;
+	}
+
+	public get containerRuntime() {
+		return this.context.containerRuntime as ContainerRuntime;
+	}
+
+	public get _runtime() {
+		return this.runtime;
+	}
+
+	public constructor(props: IDataObjectProps<I>) {
+		super(props);
+		assert((props.runtime as any).replaceChannel !== undefined, "expected migrator runtime");
+		this.hotSwapRuntime = props.runtime as UnsafeHotSwapFluidDataStoreRuntime;
+	}
+
+	// Deleting DDSes is dangerous it's best just to replace
+	public replaceChannel(channel: IChannel, factory: IChannelFactory) {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-return
+		return this.hotSwapRuntime[".UNSAFE_replaceChannel"](channel.id, factory);
+	}
+
+	public reAttachChannel(channel: IChannel) {
+		this.hotSwapRuntime[".UNSAFE_reattachChannel"](channel);
+	}
+
+	protected async initializingFirstTime(): Promise<void> {
+		const tree = SharedTree.create(this.runtime);
+		const handle: IFluidHandle<IChannel> = tree.handle as IFluidHandle<SharedTree>;
+		this.root.set("handle", handle);
+	}
+}
+
+describeNoCompat("Summarizer closes instead of refreshing", (getTestObjectProvider) => {
```

Comment:
Note to self, renaming this Hot Swap Channels, will push changes later after the pipeline turns green.